Asynchronously load aircraft, airlines and airports

tedrek opened the pull request:
This moves the airline/aircraft load to asynchronous during game play.  A demo is available at http://tedrek.github.io/atc

The tradeoff is that first load of aircraft will be a second or two later than normal, and the player needs to remain connected.  I don't think either of these are a big issue.

This should address #402.

A similiar technique could be applied to airports and geojson files if the airport list information is pre-populated.

tedrek committed:
Use asynchronous loading for aircraft

tedrek committed:
Use asynchronous loading for airlines

tedrek committed:
Use asynchronous loading for airports

tedrek commented:
With these changes the load time on my laptop is now 3 seconds, most of which is due to it being a netbook, so slow parsing.

Airports are now loaded on demand.  However the code, difficulty and name must be specified in airport.js still in order to populate the selector box.

The one thing which it doesn't do and the previous loading system did is retry in case of failure.

Any feedback would be appreciated, especially somebody trying the linked version and saying it does work for them (or things are moving twice as fast as they should be...)

eliuuk commented:
I do certainly like this better, and loads extremely quicker for me upon opening the webpage - less than a second (excluding the animations)! But, I think what can be worked on is that loading screen between  selecting a different airport, as it doesn't seem right to update the entire page with a loading screen and state what it is loading before actually showing the requested airport.

Overall, it's epic - it does the job quickly and does certainly solve the issue!

tedrek commented:
Thanks for the feedback, I'm glad it's working for you :smile: (and not just me).

Regarding the loading screen, I agree it's not the greatest (note there is no load screen for switching back to an airport which is already loaded).  Here's the issue, the game needs to pause hard while loading the new airport so the user can't interact with the canvas (this is due to the new airport not yet being in a manipulable state).  When we start the new airport loading we don't know how long it is going to take (geojson files for example can be sizeable).  It could be anywhere from 0.25 seconds where the .json is already cached and just needs to be parsed to 30+ seconds where a large .geojson needs to be downloaded by a user on dial-up.  Unfortunately I don't have any median load time statistics for our users so can't tell you what people are likely to encounter (but I encounter roughly 1 second load times).

What I wanted to avoid is somebody with a long load time sitting there seeing the old canvas but not being able to interact with it leading to the impression the game has crashed.  My feeling at the moment is that it would be best to show an animated spinner if it hasn't loaded after a delay (0.5 seconds?) and keep it up for at least 1 second (flickering screens are generally anxiety inducing).  How does that sound?

eliuuk commented:
Sounds great! It definitely seems more interesting too! :P

erikquinn commented:
This is sweet! Works really well for me too. Only thoughts I had were:
- It would be nice if they would continue trying to load all other assets in the background while it's not busy fetching files it needs for a particular operation (such as getting `aal.json` to spawn an a/c with callsign `AAL`). Then delayed spawning would only be experienced for aircraft within the first minute or two, after which everything will be fully loaded. Then we'd really be cookin'!
- small thing: I wasn't really a fan of the lowercase icao codes in the airport selection dialog... perhaps this is a result of the need for case-sensitive file names with `$.getJSON`? If so, maybe a `.toUpperCase()` could be done at the place where the icao codes are written to `$('#airport-list')`. Not a big deal, it just weirds me out to see them in lower case.

Overall, this is great and gets two thumbs up from this guy! :+1: :+1:

tedrek commented:
@erikquinn, I think this is ready for merge.  I rebased against the latest version and all airports should be present.  We'll need to be careful merging any airports in the near future and make sure they are using the new method of loading.

I've removed the loading screen as my testing showed it under 1 second for most things and I haven't had the time to implement the more complex logic, it can be implemented later without conflicts.

Regarding the display of airport ICAO codes, I didn't change anything in that area so it doesn't make sense as part of this PR.  By all means though, just create another PR for it.

Finally, for further background loading, that's a great idea but I don't have time to implement it so would prefer to merge this now to stop conflicts with new airports.  I'll try to work on the background loading as my next thing.

tedrek committed:
Remove redundant name/level from airport json

erikquinn commented:
@tedrek Cool beans, sounds good! :+1:
